srml/contract/src/lib.rs

```diff
@@ -92,6 +93,47 @@ pub trait ComputeDispatchFee<Call, Balance> {
 	fn compute_dispatch_fee(call: &Call) -> Balance;
 }
 
+#[derive(Encode,Decode,Clone,Debug)]
+/// Information for managing an acocunt and its sub trie abstraction.
+/// This is the required info to cache for an account
+pub struct AccountInfo {
+	/// unique ID for the subtree encoded as a byte
+	pub key_space: KeySpace,
+	/// the size of stored value in octet
+	pub current_mem_stored: u64,
+}
+
+/// Get a key space (key space must be unique and collision resistant depending upon its context)
+/// Note that it is different than encode because key space should have collision resistance
+/// property (being a proper uniqueid).
+pub trait KeySpaceGenerator<AccountId> {
+	/// get a keyspace for an account, using reference to parent account keyspace to ensure
+	/// uniqueness of keyspace
+	/// The implementation must ensure every new key space is unique: two consecutive call with the
+	/// same parameter needs to return different key space values.
+	fn key_space(account_id: &AccountId) -> KeySpace;
+}
+
+/// Get key space from `account_id`
+pub struct KeySpaceFromParentCounter<T: Trait>(PhantomData<T>);
+
+/// This generator use inner counter for account id and apply hash over `AccountId +
+/// parent_keyspace + accountid_counter`
+impl<T: Trait> KeySpaceGenerator<T::AccountId> for KeySpaceFromParentCounter<T>
+where
+	T::AccountId: AsRef<[u8]> {
+	fn key_space(account_id: &T::AccountId) -> KeySpace {
+    // note that skipping a value due to error is not an issue here.
+    // we only need uniqueness, not sequence.
```

Comment:
Can we change these to tabs?

```suggestion
	// we only need uniqueness, not sequence.
```